src/scanner.cpp

```diff
@@ -85,18 +100,23 @@ void Scanner::scan_number()
   float val = NAN;
   std::string str_val = source.substr(start, current - start);
   auto [ptr, ec] = std::from_chars(str_val.data(), str_val.data() + str_val.size(), val);
-  if (ec == std::errc()) { add_token(TokenType::NUMBER, val); }
+  if (ec == std::errc()) {
```

Comment:
Imagine `str_val = "123abc"`. `ec` will be `std::errc{}` because `123` is a valid `float`. You need to check the `ptr` too: it needs to be equal to the "end" pointer you passed to `from_chars`.

Reply:
Also, I wouldn't use `float`, you'll run into precision issues with large integers for example.